wip: Ensure threads and windows are cleaned up, more unit tests

rvl opened the pull request:
This is work in progress.

It uses MonadManaged and UnliftIO.bracket for resource cleanup.

One problem is that it's tricky to quickly stop the X11 `eventLoop` thread.

There are more unit tests added.

colonelpanic8 commented on src/System/Taffybar/DBus/Toggle.hs:
```diff
@@ -160,7 +160,7 @@ dbusTogglesStartupHook = do
       else
         return Nothing
-    MV.modifyMVar_ enabledVar $ const $ return $ fromMaybe M.empty mStartingMap
+    modifyMVar_ enabledVar $ const $ return $ fromMaybe M.empty mStartingMap
```
I really prefer importing into a namespace as it was here originally.

rvl replied:
OK, I have put back qualified imports of MVar.

colonelpanic8 commented:
@rvl woah how did this branch get so many conflicts? are you still working on this?

rvl commented:
A few of the conflicts would be because I split some of this PR out into #602.
Other conflicts, just because the same modules were touched by multiple PRs.
Anyway, I have rebased this branch and pushed.

This PR will be very important for reloading Taffybar within GHCi, so I intend to continue working on it.

I have been sidetracked though on building an integration test suite which will hopefully be able to find concurrency bugs. I wish to be able to reproduce issue #465.
